httpcore/_async/http11.py

```diff
@@ -281,16 +288,28 @@ def is_available(self) -> bool:
     def has_expired(self) -> bool:
         now = time.monotonic()
         keepalive_expired = self._expire_at is not None and now > self._expire_at
+        if keepalive_expired:
+            return True
 
         # If the HTTP connection is idle but the socket is readable, then the
         # only valid state is that the socket is about to return b"", indicating
         # a server-initiated disconnect.
-        server_disconnected = (
-            self._state == HTTPConnectionState.IDLE
-            and self._network_stream.get_extra_info("is_readable")
-        )
+        # Checking the readable status is relatively expensive so check it at a lower frequency.
+        if (now - self._network_stream_used_at) > self._socket_poll_interval():
+            self._network_stream_used_at = now
+            server_disconnected = (
+                self._state == HTTPConnectionState.IDLE
+                and self._network_stream.get_extra_info("is_readable")
+            )
+            if server_disconnected:
+                return True
+
+        return False
```

Comment:
Actually, I'm not happy with _interval_ calculating.

Could we improve `is_readable` instead? FWIW I noticed anyio and sync backend use `httpcore._utils.is_socket_readable` while trio backend uses its own, @MarkusSintonen is there any benchmark difference when you switch backend to trio?

For improving `is_readable` in `get_extra_info`:
- Always assume it is readable and turn it to false by specified events such as received close socket from server.
- Use synchronized Event on readability status change.

I didn't go deep to search these cases are possible or not. They are only my opinions 🤷‍♂️

Reply:
Trio also suffers badly from the constantly happening socket polling.

Without intervalled socket polling:
![old_trio](https://github.com/encode/httpcore/assets/12939780/fd9ea5da-b0f8-4636-b8c3-29929fe65301)

With intervalled socket polling:
![new_trio](https://github.com/encode/httpcore/assets/12939780/f296df23-51f4-416a-8c92-99cabecb74a9)

So in `trio` its over 5x slower when constantly doing the socket polling.

Reply:
> For improving is_readable in get_extra_info:
> Always assume it is readable and turn it to false by specified events such as received close socket from server.

Im not aware of anyway to to get events about socket getting closed. As far as I know the only way to know it is to use the socket. 🤔  But I agree the `is_readable` could be better so its not so heavy weight. We could make it just a flag based so in networking side we just set some boolean flag when we detect a network error on usage. This has a downside that we have greater probability of giving our already broken connections from the pool. But as far as I know this is how its usually done.

Reply:
Other options would be to move the intervalled polling into the network backends side. Or get even more elaborate and run the socket polling in a specific interval via `loop.call_later` which run the actual poll via `loop.run_in_executor` to avoid any possible nonasync socket IO in the async land. Gets easily hairy 😄